### PR TITLE
[Docs] yolov5_description.md in English

### PR DESCRIPTION
**Pull request by xin-li-67**
## Modification

docs/zh_cn/algorithm_descriptions/yolov5_description.md
docs/en/algorithm_descriptions/yolov5_description.md

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 1]
```diff
@@ -1 +1,189 @@
...
+MMYOLO implementation configuration: https://github.com/open-mmlab/mmyolo/blob/main/configs/yolov5/
+
+YOLOv5 official repository: https://github.com/ultralytics/yolov5
+
+## 1 v6.1 algorithm principle and MMYOLO implementation analysis
+
+YOLOv5 official release: https://github.com/ultralytics/yolov5/releases/tag/v6.1
+
+<div align=center >
+<img alt="YOLOv5精度图" src="https://user-images.githubusercontent.com/40284075/190542120-29d46b7e-ce3c-436a-9933-cfc9f86787bf.png"/>
```
Comment (PeterH0323) [inline docs/en/algorithm_descriptions/yolov5_description.md:31]:
```suggestion
<img alt="YOLOv5 accuracy" src="https://user-images.githubusercontent.com/40284075/190542120-29d46b7e-ce3c-436a-9933-cfc9f86787bf.png"/>
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 2]
```diff
@@ -1 +1,189 @@
...
+<div align=center >
+<img alt="YOLOv5精度速度图" src="https://user-images.githubusercontent.com/40284075/190542279-37734629-2b59-4bd8-a9bf-757875a93eed.png"/>
+</div>
+
+The performance is shown in the table above. YOLOv5 has two models with different scales. P6 is larger with a 1280x1280 input size, whereas P5 is the model which will be used more often. This article focuses on the structure of the P5 model.
+
+Object detection algorithms can generally be divided into data augmentation, model structure, loss calculation, etc. It is the same as YOLOv5:
+
+<div align=center >
+<img alt="训练测试策略" src="https://user-images.githubusercontent.com/40284075/190542423-f6b20d8e-c82a-4a34-9065-c161c5e29e7c.png"/>
```
Comment (PeterH0323) [inline docs/en/algorithm_descriptions/yolov5_description.md:43]:
```suggestion
<img alt="Strategy" src="https://user-images.githubusercontent.com/40284075/190542423-f6b20d8e-c82a-4a34-9065-c161c5e29e7c.png"/>
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 3]
```diff
@@ -1 +1,189 @@
...
+
+The feature maps output by the Neck module are exactly the same as the Backbone, which are (B,256,80,80), (B,512,40,40) and (B,1024,20,20).
+
+#### 1.2.3 Head
+
+The `Head` structure of YOLOv5 is exactly the same as YOLOv3, which is a `non-decoupled Head`. The Head module includes three convolution modules that do not share weights. They are used only for input feature map transformation.
+
+The `PAFPN` outputs three feature maps of different scales, whose shapes are (B,256,80,80), (B,512,40,40) and (B,1024,20,20) accordingly.
+
+Since YOLOv5 has a non-decoupled output, that is, classification and bbox detection results are all in different channels of the same convolution module. Taking the COCO 80 class as an example, when the input is 640x640 resolution, the output shapes of the Head module are (B, 3x(4+1+80),80,80), (B, 3x(4+1+80),40,40) and (B, 3x(4+1+80),20,20). 3 represents three anchors, 4 represents the bbox prediction branch, 1 represents the obj prediction branch, and 80 represents the class prediction branch.
```
Comment (PeterH0323) [inline docs/en/algorithm_descriptions/yolov5_description.md:99]:
```suggestion
Since YOLOv5 has a non-decoupled output, that is, classification and bbox detection results are all in different channels of the same convolution module. Taking the COCO 80 class as an example, when the input is 640x640 resolution, the output shapes of the Head module are `(B, 3x(4+1+80),80,80)`, `(B, 3x(4+1+80),40,40)` and `(B, 3x(4+1+80),20,20)`. 3 represents three anchors, 4 represents the bbox prediction branch, 1 represents the obj prediction branch, and 80 represents the class prediction branch.
```

**Comment by PeterH0323** [comment]
output/demo.png 不要上传文件，使用 github 链接

File: yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
+default_scope = 'mmyolo'
```
Comment (PeterH0323) [inline yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py:1]: 为啥会有这个文件？

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 5]
```diff
@@ -1 +1,639 @@
...
+</div>
+
+RangeKing@github provides the graph above. Thanks, RangeKing!
+
+YOLOv5 is an open-sourced object detection algorithm for real-time industrial applications which has received extensive attention. The reason for the explosion of YOLOv5 is not simply due to its excellent performance. It is more about the overall utility and robustness of its library.
+In short, the main features of YOLOv5 are:
+
+1. **friendly and perfect deployment supports**
+2. **fast training speech**: the training time in the case of 300 epochs is similar to most of the one-stage and two-stage algorithms under 12 epochs, such as RetinaNet, ATSS, and Faster R-CNN.
+3. **abundant optimization for corner cases**: YOLOv5 has implemented many optimizations. The functions and documentation are richer as well.
```
Comment (PeterH0323) [inline docs/en/algorithm_descriptions/yolov5_description.md:16]:
```suggestion
1. **Friendly and perfect deployment supports**
2. **Fast training speed**: the training time in the case of 300 epochs is similar to most of the one-stage and two-stage algorithms under 12 epochs, such as RetinaNet, ATSS, and Faster R-CNN.
3. **Abundant optimization for corner cases**: YOLOv5 has implemented many optimizations. The functions and documentation are richer as well.
```

File: yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py (hunk 4 above)
Comment (xin-li-67) [reply yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py:1]: > 为啥会有这个文件？

咦 我检查下我的commit记录

Comment (xin-li-67) [reply yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py]: 移除了无关文件

File: docs/zh_cn/algorithm_descriptions/yolov5_description.md [hunk 6]
```diff
@@ -255,58 +245,49 @@ train_pipeline = [
...
 
 - `Stem Layer` 是 1 个 6x6 kernel 的 `ConvModule`，相较于 v6.1 版本之前的 `Focus` 模块更加高效。
-- 前 3 个 `Stage Layer` 由 1 个 `ConvModule` 和 1 个 `CSPLayer` 组成。如上图 Details 部分，
-  其中 `ConvModule` 为 3x3 `Conv2d` + `BatchNorm` + `SiLU 激活函数`。`CSPLayer` 即 YOLOv5 官方仓库中的 C3 模块，由 3 个 `ConvModule` + n 个 `DarknetBottleneck`(带残差连接) 组成。
+- 前 3 个 `Stage Layer` 均由 1 个 `ConvModule` 和 1 个 `CSPLayer` 组成。如上图 Details 部分所示。
+  其中 `ConvModule` 为 3x3的 `Conv2d` + `BatchNorm` + `SiLU 激活函数`。`CSPLayer` 即 YOLOv5 官方仓库中的 C3 模块，由 3 个 `ConvModule` + n 个 `DarknetBottleneck`(带残差连接) 组成。
 - 第 4 个 `Stage Layer` 在最后增加了 `SPPF` 模块。`SPPF` 模块是将输入串行通过多个 5x5 大小的 `MaxPool2d` 层，与 `SPP`  模块效果相同，但速度更快。
-- P5 模型结构会在 `Stage Layer` 2-4 之后分别输出，进入 `Neck` 结构，共抽取三个输出特征图，以 640x640 输入图片为例，
-  其输出特征为 (B,256,80,80)、 (B,512,40,40) 和 (B,1024,20,20)，stride 为 8/16/32。
+- P5 模型结构会在 `Stage Layer` 2-4 之后分别输出对应结果进入 `Neck` 结构，抽取三个输出特征图，以 640x640 输入图片为例，其输出特征为 (B,256,80,80)、 (B,512,40,40) 和 (B,1024,20,20)，对应的 stride 为 8/16/32。
```
Comment (RangeKing) [inline docs/zh_cn/algorithm_descriptions/yolov5_description.md:262]:
```suggestion
- P5 模型结构会在 `Stage Layer` 2-4 之后分别输出一个特征图进入 `Neck` 结构。以 640x640 输入图片为例，其输出特征为 (B,256,80,80)、 (B,512,40,40) 和 (B,1024,20,20)，对应的 stride 分别为 8/16/32。
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 7]
```diff
@@ -1 +1,639 @@
...
+</div>
+
+The inference and post-processing include:
+
+**(1) Dimensional transformation**
+
+YOLOv5 outputs three feature maps, each is scaling at 80x80, 40x40, and 20x20. As three anchors are at each position, the output feature map channel is 3x(5+80)=255.
+YOLOv5 uses a non-decoupled output header, while most other algorithms are decoupled output headers. Therefore, to unify the post-processing logic, we decouple YOLOv5's header into the category prediction branch, bbox prediction branch, and obj prediction branch.
+
+The three scales of category prediction, bbox prediction, and obj prediction are stitched together and dimensionally transformed. For subsequent processing, the original channel dimensions are replaced at the end, and the shapes of the category prediction branch, bbox prediction branch, and obj prediction branch are (b, 3x80x80+3x40x40+3x20x20, 80)=(b,25200,80), (b,25200,4), and (b,25200,1), respectively.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:574]:
```suggestion
YOLOv5 outputs three feature maps. Each feature map is scaled at 80x80, 40x40, and 20x20. As three anchors are at each position, the output feature map channel is 3x(5+80)=255.
YOLOv5 uses a non-decoupled Head, while most other algorithms use decoupled Head. Therefore, to unify the post-processing logic, we decouple YOLOv5's Head into the category prediction branch, the bbox prediction branch, and the obj prediction branch.

The three scales of category prediction, bbox prediction, and obj prediction are stitched together and dimensionally transformed. For subsequent processing, the original channel dimensions are replaced at the end, and the shapes of the category prediction branch, bbox prediction branch, and obj prediction branch are (b, 3x80x80+3x40x40+3x20x20, 80)=(b,25200,80), (b,25200,4), and (b,25200,1), respectively.
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 8]
```diff
@@ -1 +1,639 @@
...
+
+**(2) Decoding to the original graph scale**
+
+The classification branch and obj branch need to be computed with the sigmoid function, while the bbox prediction branch needs to be decoded and reduced to the actual original graph in xyxy format.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:578]:
```suggestion
The classification branch and obj branch need to be computed with the sigmoid function, while the bbox prediction branch needs to be decoded and reduced to the original image in xyxy format.
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 9]
```diff
@@ -1 +1,639 @@
...
+            np.array(shapes) * self.img_size / self.size_divisor +
+            self.pad).astype(np.int) * self.size_divisor
+
+        for i, data_info in enumerate(data_list):
+            data_info['batch_shape'] = batch_shapes[batch_index[i]]
+```
+
+## 2 Sum up
+
+This article focuses on the principle of YOLOv5 and our implementation in MMYOLO in detail, hoping to help users understand the algorithm and the implementation process. At the same time, again, please note that since YOLOv5 itself is constantly being updated, this open-source library will also be constantly iterated. So please always check the latest version.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:639]:
```suggestion
This article focuses on the principle of YOLOv5 and our implementation in MMYOLO in detail, hoping to help users understand the algorithm and the implementation process. At the same time, again, please note that since YOLOv5 itself is constantly being updated, this open-source library will also be continuously iterated. So please always check the latest version.
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 10]
```diff
@@ -1 +1,639 @@
...
+The author adopts different weight decay strategies for different batch sizes, specifically:
+
+1. When the training batch size does not exceed 64, weight decay remains unchanged.
+2. When the training batch size exceeds 64, weight decay will be linearly scaled according to the total batch size.
+
+MMYOLO also implements through the YOLOv5OptimizerConstructor.
+
+#### 1.5.3 Gradient accumulation
+
+In order to maximize the performance under different batch sizes, the author sets the gradient accumulation function automatically when the total batch size is less than 64.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:530]:
```suggestion
To maximize the performance under different batch sizes, the author sets the gradient accumulation function automatically when the total batch size is less than 64.
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 11]
```diff
@@ -1 +1,639 @@
...
+
+The training process is similar to most YOLO, including the following strategies:
+
+1. Not using pre-trained weights.
+2. There is no multi-scale training strategy, and cudnn.benchmark can be turned on to accelerate training further.
+3. The EMA strategy is used to smooth the model.
+4. Automatic mixed-precision training with AMP by default.
+
+What needs to be reminded is that YOLOv5 officially uses single-card v100 training for the small model with a bs is 128. However, m/l/x models are trained with different numbers of multi-cards.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:539]:
```suggestion
What needs to be reminded is that the official YOLOv5 repository uses single-card v100 training for the small model with a bs is 128. However, m/l/x models are trained with different numbers of multi-cards.
```

File: docs/en/algorithm_descriptions/yolov5_description.md [hunk 12]
```diff
@@ -1 +1,639 @@
...
+Then modify the default anchor size setting in the [config file](https://github.com/open-mmlab/mmyolo/blob/main/configs/yolov5/yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py):
+
+```python
+anchors = [[(10, 13), (16, 30), (33, 23)], [(30, 61), (62, 45), (59, 119)],
+           [(116, 90), (156, 198), (373, 326)]]
+```
+
+#### 1.3.2 Bbox encoding and decoding process
+
+In anchor-based algorithms, the predicted bounding box will transform based on the pre-set anchors. Then, the transformation amount is predicted, known as the GT Bbox encoding process. Finally, the Pred Bbox decoding needs to be performed after the prediction to restore the bboxes to the original scale, known as the Pred Bbox decoding process.
```
Comment (RangeKing) [inline docs/en/algorithm_descriptions/yolov5_description.md:323]:
```suggestion
The predicted bounding box will transform based on the pre-set anchors in anchor-based algorithms. Then, the transformation amount is predicted, known as the GT Bbox encoding process. Finally, the Pred Bbox decoding needs to be performed after the prediction to restore the bboxes to the original scale, known as the Pred Bbox decoding process.
```

**Review by RangeKing — APPROVED** [review]

**Comment by RangeKing** [comment]
In chapter 1.3, “匹配策略” should be translated as `assignment strategy`, not `matching strategy`.

**Comment by xin-li-67** [comment]
> In chapter 1.3, “匹配策略” should be translated as `assignment strategy`, not `matching strategy`.

修改完毕

**Review by hhaAndroid — APPROVED** [review]
